src/permissions.rs

```diff
@@ -0,0 +1,142 @@
+        try!(Permission::validate_permission(&which_enum, &t));
+
+        // Create permission if valid
+        Ok(Permission {
+            which: which_enum,
+            target: t,
+        })
+    }
+
+    /// Validates the target for the given permission type
+    /// Returns error if invalid target
+    fn validate_permission(permission: &PermissionEnum, target: &Option<String>) -> Result<(), Error> {
+        
+        let valid = match permission {
+            &PermissionEnum::GrantPerm => {
+                target == &None::<String>
+            },
+            &PermissionEnum::EditProj => {
+                target == &None::<String>
+            },
+            &PermissionEnum::EditSeq => {
+                if target.is_none() {
+                    false
+                } else {
+                    let seq_name = target.to_owned().unwrap();
+                    let project = try!(utils::read_protonfile(None::<&Path>));
+                    project.find_sequence_by_name(&seq_name).is_some()
+                }
+            },
+            &PermissionEnum::EditSeqSec => {
+                if target.is_none() {
+                    false
+                } else {
+                    let target_str = target.to_owned().unwrap();
+                    let targets: Vec<&str> = target_str.split(",").collect();
+                    if targets.len() != 2 {
+                        false
+                    } else {
+                        let seq_name = targets[0];
+                        let section_num_str = targets[1];
+                        let section_num = match section_num_str.parse::<u32>() {
+                            Ok(n) => n,
+                            Err(_) => return Err(Error::InvalidPermissionTarget), 
+                        };
+                        let project = try!(utils::read_protonfile(None::<&Path>));
+                        match project.find_sequence_by_name(&seq_name) {
+                            Some(seq) => {
+                                section_num > 0 && section_num <= seq.num_sections
+                            },
+                            None => false,
+                        }
+
+                    }
+                }
+            },
+        };
+
+        if valid {
+            Ok(())
+        } else {
+            Err(Error::InvalidPermissionTarget)
+        }
+    }
+
+}
+
+pub fn permissions_as_string() -> String {
+    String::from("GrantPerm,EditProj,EditSeq,EditSeqSec,")
+}
+
+pub fn set_permission(
+    auth_user: &User,
+    add: bool,
+    target_username: &str,
+    permission: PermissionEnum,
+    target: Option<String>
+) -> Result<(), Error> {
+
+    // Only admins (those with GrantPerm permission) can change permissions
+    if !auth_user.is_admin() {
+        return Err(Error::UnauthorizedAction);
+    }
+
+    // Don't let users change their own permissions; that's what the admin account is for
```

Comment:
I feel like this is unnecessary. If someone's allowed to administrate the project, they should be able to change their own permissions. Like, Ian could be an admin, and say that he needs to edit a sequence.